Added astromechanics controller script and unit test

JAAlvarado-Montes opened the pull request:
Hi all,

This PR adds a new class to control an astromechanics focuser with POCS. An additional test script `test_astromech_focuser.py` was added and all the tests in there are passing when using real `astromechanics` hardware. If no hardware is found, then the tests are skipped. 

`astromechanics.py` is based off `birger.py`, with just some small differences regarding the fashion on how to send the commands and how to process their reply.

Please let me know any suggestion or comment about this PR.

Cheers,
Jaime.

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+        return None
+
+    @property
+    def max_position(self):
```
as `min_` and `max_` position don't actually return things, I think think it is best if you delete them, which I think will cause it to fall back on the base class and raise NotImplementedError: https://github.com/panoptes/POCS/blob/3f265607ce8d4427e017d6c11d3e3ebc606deea7/src/panoptes/pocs/focuser/focuser.py#L144

wtgee replied:
They are marked with `@abstractmethod` in the base class, meaning you have to override them. Probably not the best design.

wtgee replied:
@JAAlvarado-Montes @AnthonyHorton does this focuser wrap-around or something? Why no `min`/`max`? And if that's the case let's describe it in the class' docstring.

JAAlvarado-Montes replied:
Basically we *only* have three commands for the astro mechanics controller: get current position (P#), move to position XX (MXX#), and set aperture to XX (AXX#). There isn't any command to get the min/max position or additional information available for Birgers.

lspitler replied:
@JAAlvarado-Montes what happens when you specify a very big or small focus number? What focus value does it return? You could implement a quick scan of the max/min ranges by specifying really big & small moves and record the effective limits.

lspitler replied:
@JAAlvarado-Montes can you answer this question?

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+        return None
+
+    @property
+    def is_moving(self):
```
This isn't triggered during the initialize focuser.

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+    def connect(self, port):
+        try:
+            # Configure serial port.
+            # Settings copied from Bob Abraham's birger.c
```
```suggestion

```

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+
+    def _initialise(self):
+        # Get initial position of focuser adaptor.
+        self._get_initial_position()
```
I think you don't need `_get_initial_position`:
```suggestion
        self.logger.debug(f'Initial position of focuser is at {self.position} encoder units')
```

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+        """
+        self._is_moving = True
+        try:
+            ini_pos = self._send_command("P#", response_length=1)[0].replace("#", "")
```
```suggestion
            ini_pos = self.position
```

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,266 @@
+        Sends a command to the Focuser adaptor and retrieves the response.
+
+        Args:
+            command (string): command string to send (without newline), e.g. 'fa1000', 'pf'
```
```suggestion
            command (string): command string to send (without newline), e.g. "P#"
```

lspitler commented on src/panoptes/pocs/tests/test_astromech_focuser.py:
```diff
@@ -0,0 +1,88 @@
+ids = ['astromechanics']
+
+
+# Ugly hack to access id inside fixture
```
```suggestion

```

lspitler reviewed (commented):
Looks good @JAAlvarado-Montes . Will wait for @AnthonyHorton 's review if he has time. If not, let me know and I'll approve.

wtgee commented:
I’ve started a review as well, will try to finish today.

On Fri, Dec 4, 2020 at 12:12 Lee Spitler <notifications@github.com> wrote:

> *@lspitler* commented on this pull request.
>
> Looks good @JAAlvarado-Montes <https://github.com/JAAlvarado-Montes> .
> Will wait for @AnthonyHorton <https://github.com/AnthonyHorton> 's review
> if he has time. If not, let me know and I'll approve.
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536399768>:
>
> > +from contextlib import suppress
>
> +
>
> +from panoptes.pocs.focuser import AbstractFocuser
>
> +
>
> +
>
> +class Focuser(AbstractFocuser):
>
> +    """
>
> +    Focuser class for control of a Canon DSLR lens via an Astromechanics Engineering Canon EF/EF-S adapter.
>
> +
>
> +    Args:
>
> +        name (str, optional): default 'Astromechanics Focuser'
>
> +        model (str, optional): default 'Canon EF/EF-S'
>
> +        initial_position (int, optional): if given the focuser will drive to this encoder position
>
> +            following initialisation.
>
> +        dev_node_pattern (str, optional): Unix shell pattern to use to identify device nodes that
>
> +            may have a focuser adaptor attached. Default is '/dev/tty.USA49*.?', which is intended
>
>
> ⬇️ Suggested change
>
> -            may have a focuser adaptor attached. Default is '/dev/tty.USA49*.?', which is intended
>
> +            may have a focuser adaptor attached. Default is '/dev/tty.usbserial-00*'.
>
>
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536400687>:
>
> > +
>
> +from panoptes.pocs.focuser import AbstractFocuser
>
> +
>
> +
>
> +class Focuser(AbstractFocuser):
>
> +    """
>
> +    Focuser class for control of a Canon DSLR lens via an Astromechanics Engineering Canon EF/EF-S adapter.
>
> +
>
> +    Args:
>
> +        name (str, optional): default 'Astromechanics Focuser'
>
> +        model (str, optional): default 'Canon EF/EF-S'
>
> +        initial_position (int, optional): if given the focuser will drive to this encoder position
>
> +            following initialisation.
>
> +        dev_node_pattern (str, optional): Unix shell pattern to use to identify device nodes that
>
> +            may have a focuser adaptor attached. Default is '/dev/tty.USA49*.?', which is intended
>
> +            to match all the nodes created by Tripplite Keyway USA-49 USB-serial adaptors, as
>
>
> remove mention of tripplite
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536408152>:
>
> > +    Additional positonal and keyword arguments are passed to the base class, AbstractFocuser. See
>
> +    that class' documentation for a complete list.
>
> +    """
>
> +
>
> +    # Class variable to cache the device node scanning results
>
> +    _adaptor_nodes = None
>
> +
>
> +    # Class variable to store the device nodes already in use. Prevents scanning known focuser adaptors and
>
> +    # acts as a check against those ones assigned to incorrect ports.
>
> +    _assigned_nodes = []
>
> +
>
> +    def __init__(self,
>
> +                 name='Astromechanics Focuser Controller',
>
> +                 model='astromechanics',
>
> +                 initial_position=None,
>
> +                 dev_node_pattern='/dev/tty.usbserial-00*',
>
>
> would this pattern work? '/dev/tty.usbserial-00?'
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536409785>:
>
> > +    def position(self):
>
> +        """
>
> +        Returns current focus position in the lens focus encoder units.
>
> +        """
>
> +        response = self._send_command('P#', response_length=1)[0].replace("#", "")
>
> +        return response
>
> +
>
> +    @property
>
> +    def min_position(self):
>
> +        """
>
> +        Returns position of close limit of focus travel, in encoder units.
>
> +        """
>
> +        return None
>
> +
>
> +    @property
>
> +    def max_position(self):
>
>
> as min_ and max_ position don't actually return things, I think think it
> is best if you delete them, which I think will cause it to fall back on the
> base class and raise NotImplementedError:
> https://github.com/panoptes/POCS/blob/3f265607ce8d4427e017d6c11d3e3ebc606deea7/src/panoptes/pocs/focuser/focuser.py#L144
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536410120>:
>
> > +    @property
>
> +    def min_position(self):
>
> +        """
>
> +        Returns position of close limit of focus travel, in encoder units.
>
> +        """
>
> +        return None
>
> +
>
> +    @property
>
> +    def max_position(self):
>
> +        """
>
> +        Returns position of far limit of focus travel, in encoder units.
>
> +        """
>
> +        return None
>
> +
>
> +    @property
>
> +    def is_moving(self):
>
>
> This isn't triggered during the initialize focuser.
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536410203>:
>
> > +        """
>
> +        return None
>
> +
>
> +    @property
>
> +    def is_moving(self):
>
> +        """ True if the focuser is currently moving. """
>
> +        return self._is_moving
>
> +
>
> +    ##################################################################################################
>
> +    # Public Methods
>
> +    ##################################################################################################
>
> +
>
> +    def connect(self, port):
>
> +        try:
>
> +            # Configure serial port.
>
> +            # Settings copied from Bob Abraham's birger.c
>
>
> ⬇️ Suggested change
>
> -            # Settings copied from Bob Abraham's birger.c
>
>
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536411148>:
>
> > +                response.append(self._serial_io.readline())
>
> +
>
> +        elif response_length > 0:
>
> +            # Expecting some number of lines of response. Attempt to read that many lines.
>
> +            for i in range(response_length):
>
> +                response.append(self._serial_io.readline())
>
> +
>
> +        else:
>
> +            # Don't know what to expect. Call readlines() to get whatever is there.
>
> +            response.append(self._serial_io.readlines())
>
> +
>
> +        return response
>
> +
>
> +    def _initialise(self):
>
> +        # Get initial position of focuser adaptor.
>
> +        self._get_initial_position()
>
>
> I think you don't need _get_initial_position:
> ⬇️ Suggested change
>
> -        self._get_initial_position()
>
> +        self.logger.debug(f'Initial position of focuser is at {self.position} encoder units')
>
>
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536411358>:
>
> > +    def move_by(self, increment):
>
> +        """
>
> +        Move focuser by a given amount.
>
> +
>
> +        Args:
>
> +            increment (int): distance to move the focuser, in encoder units.
>
> +
>
> +        Returns:
>
> +            int: distance moved, in encoder units.
>
> +
>
> +        Does not do any checking of the requested increment but will warn if the lens reports
>
> +        hitting a stop.
>
> +        """
>
> +        self._is_moving = True
>
> +        try:
>
> +            ini_pos = self._send_command("P#", response_length=1)[0].replace("#", "")
>
>
> ⬇️ Suggested change
>
> -            ini_pos = self._send_command("P#", response_length=1)[0].replace("#", "")
>
> +            ini_pos = self.position
>
>
> ------------------------------
>
> In src/panoptes/pocs/focuser/astromechanics.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536412164>:
>
> > +            # returned then the focuser is no longer moving.
>
> +            self._is_moving = False
>
> +
>
> +        self.logger.debug("Moved by {} encoder units. Current position is {}".format(increment, new_pos))
>
> +        return new_pos
>
> +
>
> +    ##################################################################################################
>
> +    # Private Methods
>
> +    ##################################################################################################
>
> +
>
> +    def _send_command(self, command, response_length=None):
>
> +        """
>
> +        Sends a command to the Focuser adaptor and retrieves the response.
>
> +
>
> +        Args:
>
> +            command (string): command string to send (without newline), e.g. 'fa1000', 'pf'
>
>
> ⬇️ Suggested change
>
> -            command (string): command string to send (without newline), e.g. 'fa1000', 'pf'
>
> +            command (string): command string to send (without newline), e.g. "P#"
>
>
> ------------------------------
>
> In src/panoptes/pocs/tests/test_astromech_focuser.py
> <https://github.com/panoptes/POCS/pull/1041#discussion_r536412723>:
>
> > @@ -0,0 +1,88 @@
>
> +import time
>
> +
>
> +import pytest
>
> +from threading import Thread
>
> +
>
> +from panoptes.utils.config.helpers import load_config
>
> +
>
> +from panoptes.pocs.focuser.astromechanics import Focuser as AstroMechFocuser
>
> +
>
> +params = [AstroMechFocuser]
>
> +ids = ['astromechanics']
>
> +
>
> +
>
> +# Ugly hack to access id inside fixture
>
>
> ⬇️ Suggested change
>
> -# Ugly hack to access id inside fixture
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/panoptes/POCS/pull/1041#pullrequestreview-545321808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAARIUEX7I25GBSMWBJGODTSTFNEHANCNFSM4UM2ZGQA>
> .
>
-- 
~Wilfred Tyler Gee

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,263 @@
+    def __init__(self,
+                 name='Astromechanics Focuser Controller',
+                 model='astromechanics',
+                 initial_position=None,
+                 dev_node_pattern='/dev/tty.usbserial-00*',
```
Neither the `initial_position` or the `dev_node_pattern` are actually used anywhere.

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,263 @@
+        with suppress(AttributeError):
+            device_node = self.port
+            Focuser._assigned_nodes.remove(device_node)
+            self.logger.debug('Removed {} from assigned nodes list'.fomat(device_node))
```
```suggestion
            self.logger.debug('Removed {} from assigned nodes list'.format(device_node))
```

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,263 @@
+        Returns current focus position in the lens focus encoder units.
+        """
+        response = self._send_command('P#', response_length=1)[0].replace("#", "")
+        return response
```
The other focusers are returning ints here.

```suggestion
        return int(response)
```

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,263 @@
+        """
+        self._is_moving = True
+        try:
+            self._send_command('M{:d}#'.format(int(new_position)), response_length=0)
```
@JAAlvarado-Montes, I think the f-strings are a lot cleaner in most places and would encourage you to use them as the default, but especially in something like this I think the extra `format` call makes it harder to see the command being written.

```suggestion
            self._send_command(f'M{int(new_position):d}#'), response_length=0)
```

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,263 @@
+        Returns:
+            int: focuser position following the move, in encoder units.
+
+        Does not do any checking of the requested position but will warn if the lens reports
+        hitting a stop.
```
Is this true for this focuser?

lspitler replied:
@JAAlvarado-Montes can you address this? It does not warn at present.

wtgee reviewed (commented):
@JAAlvarado-Montes why not add this focuser to the existing focuser tests rather than create a whole new separate file?

In addition to that, there seems to be a lot of duplication with the Birger class. I realize that's because it's just being copied, but if we are going to have both of them we should probably abstract things out a big more.  For the mounts, there is an `AbstractMount` and then an `AbstractSerialMount` to handle all the shared serial functions, which is a lot of the duplication here.

JAAlvarado-Montes commented:
Hi @lspitler @wtgee @AnthonyHorton ,

I have addressed all of your comments so this PR is ready for you to take a look at it again. Basically, min/max methods can't be removed @lspitler because they are marked with the decorator `@abstractmethod` so we have to override them. 

Also, @wtgee, I decided to have a separate test file for the astromechanics focuser because they behave quite differently to the Birgers and many tests in `test_focuser.py` won't work for the astromechanics hardware in the way they're written. I didn't understand very well what you said about using `AbstractMount` and `AbstractSerialMount`, since the two scripts on this PR do not use anything from that, and basically there are only three methods given that these focusers only have three commands available.

PS. I already tested this with real hardware and it's working pretty well. Once this PR is merged we'll be able to test the telescope with one of the lenses having an astromechanics focuser controller.

Cheers,
Jaime.

lspitler commented:
@JAAlvarado-Montes I think the comments about abstract mounts is a way of saying to make another class that sits between `AbstractFocuser` and `Focuser` from Birger or AstroMech. So a `AbstractSerialFocuser` that captures all the duplicated code for a serial-driven Focuser device (which is both the Birger and AstroMechs). Does this make sense?

JAAlvarado-Montes commented:
Kindda @lspitler, doesn't that make things more complicated if the class for astro mechanics already has what it needs?

lspitler commented:
Problem @JAAlvarado-Montes is that we now have 2 versions of nearly identical code. I can imagine you'll improve AstroMech someday and maybe the improvement would also improve the birger serial code as the function is identical. The proposed setup means any future fix or upgrade immediately benefits birger or astromech devices.

JAAlvarado-Montes commented:
Hmm but what I'd put into that new class? They are not completely identical, I had to change various things inside the `Focuser class` in `astromechanics.py` to make the focuser work (mostly the commands). Also, the `__init__` function of `Focuser` is slightly different in astromechanics.py and in birger.py, as now we neither have `dev_node_pattern` nor `initial_position`

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,165 @@
+        super().__init__(name=name, model=model, *args, **kwargs)
+        self.logger.debug('Initialising Astromechanics Lens Controller')
+
+        self._serial_port.baudrate = 38400
```
Don't think this will work as `self._serial_port` is created only when `_connect` is run, which is hard-wired to `self._serial_port.baudrate = 115200`. I guess you haven't tested this code on the hardware?

My hope with adding this line was to move `def _connect(self, port):` into `serial.py` as it is mostly the same code.

You could move it there and add a required arg of `baudrate`, so: `def _connect(self, port, baudrate):`.

And then call it from birger or astromech like: `super().__init__(name=name, model=model, baudrate=38400, *args, **kwargs)` in `__init__` of `astromechanic.py`. 

Then modify `serial.py` to call `self.connect(self.port, baudrate)`

Be sure to delete `_connect` definition from `birger.py`

JAAlvarado-Montes replied:
It is working with the astro mechanics hardware. I was able to move a small lens and one of the Huntsman lenses.

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,165 @@
+    # Private Methods
+    ##################################################################################################
+
+    def _send_command(self, command, response_length=None):
```
See my suggestions elsewhere in this file to make this happen on each call, e.g.: `self._send_command('A00#', ignore_response=True)`

lspitler commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,165 @@
+    def _initialise_aperture(self):
+        self.logger.debug('Initialising aperture motor')
+        self._send_command('A00#')
+        self.logger.debug('Aperture initialised at maximum value')
```
Is this more correct?

```suggestion
        self.logger.debug('Aperture initialised')
```

lspitler commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -63,19 +62,16 @@ class Focuser(AbstractFocuser):
     # Class variable to cache the device node scanning results
     _birger_nodes = None
```
```suggestion
```

Don't need this as `_adaptor_nodes = None` in `serial.py` serves this purpose more generically.

lspitler replied:
Change all use of `_birger_nodes` to `_adaptor_nodes`.

lspitler reviewed (changes requested):
@JAAlvarado-Montes some more important comments - can you also make sure to test your code on the birger and astromechanic hardware after the modifications & unit tests pass?

lspitler commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -0,0 +1,99 @@
+            connected = self._serial_port.isOpen()
+        return connected
+
+    def _connect(self, port):
```
```suggestion
    def _connect(self, port, baudrate):
```

lspitler commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -0,0 +1,99 @@
+            # Settings copied from Bob Abraham's birger.c
+            self._serial_port = serial.Serial()
+            self._serial_port.port = port
+            self._serial_port.baudrate = 115200
```
```suggestion
            self._serial_port.baudrate = baudrate
```

lspitler commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -0,0 +1,99 @@
+    def _connect(self, port):
+        try:
+            # Configure serial port.
+            # Settings copied from Bob Abraham's birger.c
```
```suggestion
```

JAAlvarado-Montes commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -0,0 +1,165 @@
+        """
+        Returns current focus position in the lens focus encoder units.
+        """
+        response = self._send_command('P#').rstrip("#")
```
Can't use ignore_response here cause then we wouldn't have the position. "P#" is actually the only command of astromechanics which sends back a reply.

JAAlvarado-Montes commented:
Hi @lspitler @wtgee @AnthonyHorton ,

After a long debugging process with the creation of the `AbstractSerialFocuser` class, I've managed to reuse (serialize) all the common code that was duplicated in both `astromechanics.py` and `birger.py`.  The new `serial.py` contains all the general and common properties for both focuser devices (i.e. birger and astromechanics), and all the particular methods and properties for each device are contained in its corresponding class.

The additional script that tested only astromechanics.py was removed and now the same tests are run for both astromechanics and birger devices. The good news is that unit tests are passing when using either hardware, and all the changes have been tested by moving a small DSLR lens and one of the big Huntsman lenses.

That said, when you have some time please take a look at this PR, thanks in advance.

Cheers,
Jaime.

lspitler reviewed (approved):
@JAAlvarado-Montes looks really good! Well done.

Can you please remove 3 mentions of "birger" in `serial.py`? Make them more generic?

I think you can remove `ignore_response` in `serial.py` as you now have a special case for astromechs.

ALso 
`            response = self._serial_io.readline()`
is the same as:
`        echo = self._serial_io.readline().rstrip()`
just a few lines later.

Can you move it out and just get rid of the use of `echo`?

So `error_pattern` and `error_messages` are specific to birger, so they should live in `birger.py`. Can you add these as optional variables as arguments to `_send_command` that `birger.py` sends? OR maybe just move them to `birger.py` and the namespace will mean it will just work?

Once the above are finished, happy for you to merge if you can. Do let me know if you have questions on any point.

 @wtgee might need to do the merge?

JAAlvarado-Montes commented:
Hi @lspitler ,

I've addressed all the points above. Just a thing: `response = self._serial_io.readline()` and `response = self._serial_io.readline().rstrip()` are not exactly the same as they depend on how the hardware sends the reply, so in the case of the birgers all the replies are sent with an empty trailing space, whereas `astromechanics` sends the replies with a "#". So if I did what you suggested above (which I did at first), the commands are not sent to the birger. For birgers we have to use the `.rstrip()` at the end of `self._serial_io.readline()`.

That said, everything else has been done. Now `error_pattern` and `error_messages` are optional arguments of `_send_command` and they were removed from `serial.py`. Now `serial.py` is completely generic stuff.

Cheers,
Jaime.

wtgee commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -0,0 +1,180 @@
+        # Send command
+        self._serial_io.write(command + '\r')
+
+        if self.model == "astromechanics":
+            return self._serial_io.readline()
```
Is everything below specific to a birger?  Seems like the more generic case is to just do the read and return and the more specific case should be handled where it is needed.

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -308,78 +233,6 @@ def move_by(self, increment):
     # Private Methods
     ##################################################################################################
 
-    def _send_command(self, command, response_length=None, ignore_response=False):
```
If the birger needs these error messages for call to `_send_command` then you can just override the method here from the base class, then you don't need to put the parameters in the call each time.

See also the note about the birger specific commands in the base class, which could just be moved back here.

wtgee reviewed (approved):
This looks good, although I think the birger specific items in `_send_command` need to be moved back to the birger class. Let me know if any questions about that.
